Is AWS::Include and Fn::Transform implemented for CloudFormation?

Hello,

I am curious as to whether or not AWS::Include and Fn::Transform are implemented in the CloudFormation mock?

moto v1.3.8 installed via pip on MacOS using pytest
boto3 version 1.9.160
platform darwin -- Python 3.7.3, pytest-4.5.0, py-1.8.0, pluggy-0.12.0

Here's my test, please let me know if I'm doing anything wrong:

```
from unittest import TestCase
import moto
import boto3
import json

@moto.mock_cloudformation
@moto.mock_s3
def test_launch_ec2_cluster(self):
    s3_resource = boto3.resource('s3')
    s3_resource.create_bucket(Bucket='owi-common-cf', CreateBucketConfiguration={
        'LocationConstraint': 'us-west-2'})
    s3_resource.Object('owi-common-cf', 'snippets/test.json').put(Body=json.dumps({
        'lgname': {
            'name': 'some-log-group'
        }
    }))
    template = {
        "AWSTemplateFormatVersion": "2010-09-09",
        "Mappings": {
            "EnvironmentMapping": {
                "Fn::Transform": {
                    "Name": "AWS::Include",
                    "Parameters": {
                        "Location": "s3://owi-common-cf/snippets/test.json"
                    }
                }
            }
        },
        "Resources": {
            "LogGroup": {
                "Properties": {
                    "LogGroupName": {
                        "Fn::FindInMap": [
                            "EnvironmentMapping",
                            "lgname",
                            "name"
                        ]
                    },
                    "RetentionInDays": 90
                },
                "Type": "AWS::Logs::LogGroup"
            }
        }
    }

    cf_conn = boto3.client('cloudformation', 'us-west-2')
    cf_conn.create_stack(
        StackName="test_stack",
        TemplateBody=json.dumps(template),
    )

    self.assertTrue(True)

```

The error shows up as:
```
=================================== FAILURES ===================================
____________________ TestEc2Cluster.test_launch_ec2_cluster ____________________

self = <python.tests.test_ec2_cluster.TestEc2Cluster testMethod=test_launch_ec2_cluster>

    @moto.mock_cloudformation
    @moto.mock_s3
    def test_launch_ec2_cluster(self):
        s3_resource = boto3.resource('s3')
        s3_resource.create_bucket(Bucket='owi-common-cf', CreateBucketConfiguration={
            'LocationConstraint': 'us-west-2'})
        s3_resource.Object('owi-common-cf', 'snippets/test.json').put(Body=json.dumps({
            'lgname': {
                'name': 'some-log-group'
            }
        }))
        template = {
            "AWSTemplateFormatVersion": "2010-09-09",
            "Mappings": {
                "EnvironmentMapping": {
                    "Fn::Transform": {
                        "Name": "AWS::Include",
                        "Parameters": {
                            "Location": {
                                "Fn::Sub": "s3://owi-common-cf/snippets/test.json"
                            }
                        }
                    }
                }
            },
            "Resources": {
                "LogGroup": {
                    "Properties": {
                        "LogGroupName": {
                            "Fn::FindInMap": [
                                "EnvironmentMapping",
                                "lgname",
                                "name"
                            ]
                        },
                        "RetentionInDays": 90
                    },
                    "Type": "AWS::Logs::LogGroup"
                }
            }
        }
    
        cf_conn = boto3.client('cloudformation', 'us-west-2')
        cf_conn.create_stack(
            StackName="test_stack",
>           TemplateBody=json.dumps(template),
        )

resources/gov/usgs/wma/python/libs/python/tests/test_ec2_cluster.py:88: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.7/site-packages/botocore/client.py:357: in _api_call
    return self._make_api_call(operation_name, kwargs)
/usr/local/lib/python3.7/site-packages/botocore/client.py:648: in _make_api_call
    operation_model, request_dict, request_context)
/usr/local/lib/python3.7/site-packages/botocore/client.py:667: in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
/usr/local/lib/python3.7/site-packages/botocore/endpoint.py:102: in make_request
    return self._send_request(request_dict, operation_model)
/usr/local/lib/python3.7/site-packages/botocore/endpoint.py:137: in _send_request
    success_response, exception):
/usr/local/lib/python3.7/site-packages/botocore/endpoint.py:231: in _needs_retry
    caught_exception=caught_exception, request_dict=request_dict)
/usr/local/lib/python3.7/site-packages/botocore/hooks.py:356: in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
/usr/local/lib/python3.7/site-packages/botocore/hooks.py:228: in emit
    return self._emit(event_name, kwargs)
/usr/local/lib/python3.7/site-packages/botocore/hooks.py:211: in _emit
    response = handler(**kwargs)
/usr/local/lib/python3.7/site-packages/botocore/retryhandler.py:183: in __call__
    if self._checker(attempts, response, caught_exception):
/usr/local/lib/python3.7/site-packages/botocore/retryhandler.py:251: in __call__
    caught_exception)
/usr/local/lib/python3.7/site-packages/botocore/retryhandler.py:269: in _should_retry
    return self._checker(attempt_number, response, caught_exception)
/usr/local/lib/python3.7/site-packages/botocore/retryhandler.py:317: in __call__
    caught_exception)
/usr/local/lib/python3.7/site-packages/botocore/retryhandler.py:223: in __call__
    attempt_number, caught_exception)
/usr/local/lib/python3.7/site-packages/botocore/retryhandler.py:359: in _check_caught_exception
    raise caught_exception
/usr/local/lib/python3.7/site-packages/botocore/endpoint.py:197: in _do_get_response
    responses = self._event_emitter.emit(event_name, request=request)
/usr/local/lib/python3.7/site-packages/botocore/hooks.py:356: in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
/usr/local/lib/python3.7/site-packages/botocore/hooks.py:228: in emit
    return self._emit(event_name, kwargs)
/usr/local/lib/python3.7/site-packages/botocore/hooks.py:211: in _emit
    response = handler(**kwargs)
/usr/local/lib/python3.7/site-packages/moto/core/models.py:298: in __call__
    status, headers, body = response_callback(request, request.url, request.headers)
/usr/local/lib/python3.7/site-packages/moto/core/responses.py:117: in dispatch
    return cls()._dispatch(*args, **kwargs)
/usr/local/lib/python3.7/site-packages/moto/core/responses.py:200: in _dispatch
    return self.call_action()
/usr/local/lib/python3.7/site-packages/moto/core/responses.py:274: in call_action
    response = method()
/usr/local/lib/python3.7/site-packages/moto/cloudformation/responses.py:68: in create_stack
    role_arn=role_arn,
/usr/local/lib/python3.7/site-packages/moto/cloudformation/models.py:414: in create_stack
    create_change_set=create_change_set,
/usr/local/lib/python3.7/site-packages/moto/cloudformation/models.py:187: in __init__
    self.resource_map = self._create_resource_map()
/usr/local/lib/python3.7/site-packages/moto/cloudformation/models.py:195: in _create_resource_map
    resource_map.create()
/usr/local/lib/python3.7/site-packages/moto/cloudformation/parsing.py:463: in create
    if isinstance(self[resource], ec2_models.TaggedEC2Resource):
/usr/local/lib/python3.7/site-packages/moto/cloudformation/parsing.py:403: in __getitem__
    resource_logical_id, resource_json, self, self._region_name)
/usr/local/lib/python3.7/site-packages/moto/cloudformation/parsing.py:294: in parse_and_create_resource
    resource_tuple = parse_resource(logical_id, resource_json, resources_map)
/usr/local/lib/python3.7/site-packages/moto/cloudformation/parsing.py:272: in parse_resource
    resource_json = clean_json(resource_json, resources_map)
/usr/local/lib/python3.7/site-packages/moto/cloudformation/parsing.py:223: in clean_json
    cleaned_val = clean_json(value, resources_map)
/usr/local/lib/python3.7/site-packages/moto/cloudformation/parsing.py:223: in clean_json
    cleaned_val = clean_json(value, resources_map)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

resource_json = {'Fn::FindInMap': ['EnvironmentMapping', 'lgname', 'name']}
resources_map = <moto.cloudformation.parsing.ResourceMap object at 0x125167080>

    def clean_json(resource_json, resources_map):
        """
        Cleanup the a resource dict. For now, this just means replacing any Ref node
        with the corresponding physical_resource_id.
    
        Eventually, this is where we would add things like function parsing (fn::)
        """
        if isinstance(resource_json, dict):
            if 'Ref' in resource_json:
                # Parse resource reference
                resource = resources_map[resource_json['Ref']]
                if hasattr(resource, 'physical_resource_id'):
                    return resource.physical_resource_id
                else:
                    return resource
    
            if "Fn::FindInMap" in resource_json:
                map_name = resource_json["Fn::FindInMap"][0]
                map_path = resource_json["Fn::FindInMap"][1:]
                result = resources_map[map_name]
                for path in map_path:
>                   result = result[clean_json(path, resources_map)]
E                   KeyError: 'lgname'

/usr/local/lib/python3.7/site-packages/moto/cloudformation/parsing.py:148: KeyError
```

--- a/moto/cloudformation/parsing.py
+++ b/moto/cloudformation/parsing.py
@@ -1,5 +1,6 @@
 from __future__ import unicode_literals
 import functools
+import json
 import logging
 import copy
 import warnings
@@ -24,7 +25,8 @@ from moto.rds import models as rds_models
 from moto.rds2 import models as rds2_models
 from moto.redshift import models as redshift_models
 from moto.route53 import models as route53_models
-from moto.s3 import models as s3_models
+from moto.s3 import models as s3_models, s3_backend
+from moto.s3.utils import bucket_and_name_from_url
 from moto.sns import models as sns_models
 from moto.sqs import models as sqs_models
 from moto.core import ACCOUNT_ID
@@ -150,7 +152,10 @@ def clean_json(resource_json, resources_map):
             map_path = resource_json["Fn::FindInMap"][1:]
             result = resources_map[map_name]
             for path in map_path:
-                result = result[clean_json(path, resources_map)]
+                if "Fn::Transform" in result:
+                    result = resources_map[clean_json(path, resources_map)]
+                else:
+                    result = result[clean_json(path, resources_map)]
             return result
 
         if "Fn::GetAtt" in resource_json:
@@ -470,6 +475,17 @@ class ResourceMap(collections_abc.Mapping):
     def load_mapping(self):
         self._parsed_resources.update(self._template.get("Mappings", {}))
 
+    def transform_mapping(self):
+        for k, v in self._template.get("Mappings", {}).items():
+            if "Fn::Transform" in v:
+                name = v["Fn::Transform"]["Name"]
+                params = v["Fn::Transform"]["Parameters"]
+                if name == "AWS::Include":
+                    location = params["Location"]
+                    bucket_name, name = bucket_and_name_from_url(location)
+                    key = s3_backend.get_key(bucket_name, name)
+                    self._parsed_resources.update(json.loads(key.value))
+
     def load_parameters(self):
         parameter_slots = self._template.get("Parameters", {})
         for parameter_name, parameter in parameter_slots.items():
@@ -515,6 +531,7 @@ class ResourceMap(collections_abc.Mapping):
 
     def create(self):
         self.load_mapping()
+        self.transform_mapping()
         self.load_parameters()
         self.load_conditions()
 

--- a/moto/cloudwatch/models.py
+++ b/moto/cloudwatch/models.py
@@ -5,6 +5,7 @@ from boto3 import Session
 from moto.core.utils import iso_8601_datetime_without_milliseconds
 from moto.core import BaseBackend, BaseModel
 from moto.core.exceptions import RESTError
+from moto.logs import logs_backends
 from datetime import datetime, timedelta
 from dateutil.tz import tzutc
 from uuid import uuid4
@@ -428,12 +429,9 @@ class LogGroup(BaseModel):
         cls, resource_name, cloudformation_json, region_name
     ):
         properties = cloudformation_json["Properties"]
-        spec = {"LogGroupName": properties["LogGroupName"]}
-        optional_properties = "Tags".split()
-        for prop in optional_properties:
-            if prop in properties:
-                spec[prop] = properties[prop]
-        return LogGroup(spec)
+        log_group_name = properties["LogGroupName"]
+        tags = properties.get("Tags", {})
+        return logs_backends[region_name].create_log_group(log_group_name, tags)
 
 
 cloudwatch_backends = {}

--- a/moto/logs/models.py
+++ b/moto/logs/models.py
@@ -405,6 +405,7 @@ class LogsBackend(BaseBackend):
         if log_group_name in self.groups:
             raise ResourceAlreadyExistsException()
         self.groups[log_group_name] = LogGroup(self.region_name, log_group_name, tags)
+        return self.groups[log_group_name]
 
     def ensure_log_group(self, log_group_name, tags):
         if log_group_name in self.groups:

--- a/moto/s3/utils.py
+++ b/moto/s3/utils.py
@@ -35,6 +35,17 @@ def bucket_name_from_url(url):
             return None
 
 
+# 'owi-common-cf', 'snippets/test.json' = bucket_and_name_from_url('s3://owi-common-cf/snippets/test.json')
+def bucket_and_name_from_url(url):
+    prefix = "s3://"
+    if url.startswith(prefix):
+        bucket_name = url[len(prefix) : url.index("/", len(prefix))]
+        key = url[url.index("/", len(prefix)) + 1 :]
+        return bucket_name, key
+    else:
+        return None, None
+
+
 REGION_URL_REGEX = re.compile(
     r"^https?://(s3[-\.](?P<region1>.+)\.amazonaws\.com/(.+)|"
     r"(.+)\.s3[-\.](?P<region2>.+)\.amazonaws\.com)/?"

--- a/tests/test_cloudformation/test_cloudformation_stack_integration.py
+++ b/tests/test_cloudformation/test_cloudformation_stack_integration.py
@@ -32,12 +32,14 @@ from moto import (
     mock_iam_deprecated,
     mock_kms,
     mock_lambda,
+    mock_logs,
     mock_rds_deprecated,
     mock_rds2,
     mock_rds2_deprecated,
     mock_redshift,
     mock_redshift_deprecated,
     mock_route53_deprecated,
+    mock_s3,
     mock_sns_deprecated,
     mock_sqs,
     mock_sqs_deprecated,
@@ -2332,3 +2334,48 @@ def test_stack_dynamodb_resources_integration():
     response["Item"]["Sales"].should.equal(Decimal("10"))
     response["Item"]["NumberOfSongs"].should.equal(Decimal("5"))
     response["Item"]["Album"].should.equal("myAlbum")
+
+
+@mock_cloudformation
+@mock_logs
+@mock_s3
+def test_create_log_group_using_fntransform():
+    s3_resource = boto3.resource("s3")
+    s3_resource.create_bucket(
+        Bucket="owi-common-cf",
+        CreateBucketConfiguration={"LocationConstraint": "us-west-2"},
+    )
+    s3_resource.Object("owi-common-cf", "snippets/test.json").put(
+        Body=json.dumps({"lgname": {"name": "some-log-group"}})
+    )
+    template = {
+        "AWSTemplateFormatVersion": "2010-09-09",
+        "Mappings": {
+            "EnvironmentMapping": {
+                "Fn::Transform": {
+                    "Name": "AWS::Include",
+                    "Parameters": {"Location": "s3://owi-common-cf/snippets/test.json"},
+                }
+            }
+        },
+        "Resources": {
+            "LogGroup": {
+                "Properties": {
+                    "LogGroupName": {
+                        "Fn::FindInMap": ["EnvironmentMapping", "lgname", "name"]
+                    },
+                    "RetentionInDays": 90,
+                },
+                "Type": "AWS::Logs::LogGroup",
+            }
+        },
+    }
+
+    cf_conn = boto3.client("cloudformation", "us-west-2")
+    cf_conn.create_stack(
+        StackName="test_stack", TemplateBody=json.dumps(template),
+    )
+
+    logs_conn = boto3.client("logs", region_name="us-west-2")
+    log_group = logs_conn.describe_log_groups()["logGroups"][0]
+    log_group["logGroupName"].should.equal("some-log-group")

